### ISSUE TITLE
Finally blocks cause unnecessary stack spilling

### ISSUE THREAD

**Issue by hez2010** [dotnet/runtime#65658]
### Description
Given this code:
```csharp
public class C
{
    public int M1(int x)
    {
        try
        {
            x++;
        }
        finally
        {
            x++;
        }
        return x;
    }
    
    public int M2(int x)
    {
        try
        {
            x++;
        }
        finally
        {
        }
        return x;
    }
    
    public int M3(int x)
    {
        try
        {
        }
        finally
        {
            x++;
        }
        return x;
    }
}
```

With x86 JIT, the `x` in `M1` and `M3` will be spilled to stack, while `M2` won't.

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQMwAJMBMuAwgLABQA3hbrfnrArgLICMAFI7gB4CUNdauTojcCAE4BPAaNxDZs7gGolAbhmiAvhpEAzWBAA2h6cIXyFI5Wp11tZ2ZgDsPdQ9z2RGnLi7MCnDBMfBoWshKmlmGWPCpulp4K+jBGJra00SKJji7c8XYa3gxBLNiBwfzumXQR6XJ12aLJqZHmdVZxDXXOrhr2mkA

With x64 JIT, only `x` in `M1` will be spilled to stack.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAZgAJiAmcgYQFgAoAbyfPcooEsA7DcgLIBGABS9+CAJRsOrRhwXkMUAJ4zF5ORo0IA1LoDc6xQF9jCgGa9sAGxtr52rdoV7D5jmccbiAdnIIRt7kXgrGZOTiglRifAHSwc4ayg4uSS4B+kEuodpWPLb2HuzpCrk+/oHG5eTh3HECpLESCQqlHCnFml01ltZ2qU5drlk9XX4B2Z5MJkA=

### Configuration
.NET 6

**Comment by EgorBo** [comment]
It's not about spilling more like JIT is able to drop empty try {} and finally {} blocks and avoid all the EH overhead.
But it's not able to do that for arbitrary never-throw blocks (at least in unchecked context for this case)

**Comment by hez2010** [comment]
> JIT is able to drop empty try {} and finally {} blocks

Why is the empty try{} block not being dropped on x86 JIT?

**Comment by EgorBo** [comment]
> > JIT is able to drop empty try {} and finally {} blocks
> 
> Why is the empty try{} block not being dropped on x86 JIT?

Judging by the JITDUMP we drop it, but then on x86 we still have EHSlots with must-init
```
; Final local variable assignments
;
;* V00 this         [V00    ] (  0,  0   )     ref  ->  zero-ref    this class-hnd single-def
;  V01 arg1         [V01,T00] (  5,  5   )     int  ->  edx        
;  V02 EHSlots      [V02    ] (  1,  1   )     blk (16) [ebp-14H]   must-init "lvaShadowSPslotsVar"
;

Found 4 lvMustInit int-sized stack slots, frame offsets 20 through 4
IN0003:        push     ebp
IN0004:        mov      ebp, esp
IN0005:        sub      esp, 20
IN0006:        xor      eax, eax
IN0007:        mov      dword ptr [V02 ebp-14H], eax
IN0008:        mov      dword ptr [V02+0x4 ebp-10H], eax
IN0009:        mov      dword ptr [V02+0x8 ebp-0CH], eax
IN000a:        mov      dword ptr [V02+0xC ebp-08H], eax
```
Sounds like something not difficult to fix, although we generally focus more on 64bit performance, so 32bit one often is left with lower priority

cc @dotnet/jit-contrib @SingleAccretion

**Comment by EgorBo** [comment]
Looks like `compXcptnsCount` is not updated after we droped try-finally, let me see

**Comment by AndyAyersMS** [comment]
Verified fixed on current main.

Running the same repro through the x86 cross-JIT with `JitDisasm` for `M1`/`M2`/`M3`:

```
; Assembly listing for method C:M1(int):int:this (FullOpts)
; esp based frame
;  V01 arg1         [V01,T00] (  3,  3   )     int  ->  edx         single-def
; Lcl frame size = 0
       lea      eax, [edx+0x02]
       ret

; Assembly listing for method C:M2(int):int:this (FullOpts)
; esp based frame
;  V01 arg1         [V01,T00] (  3,  3   )     int  ->  edx         single-def
; Lcl frame size = 0
       lea      eax, [edx+0x01]
       ret

; Assembly listing for method C:M3(int):int:this (FullOpts)
; esp based frame
;  V01 arg1         [V01,T00] (  3,  3   )     int  ->  edx         single-def
; Lcl frame size = 0
       lea      eax, [edx+0x01]
       ret
```

All three are 4 bytes, esp-based, with `x` living in `edx` -- the same shape as x64.

Three pieces of work contributed:
- #108003 added a late pass of empty try/finally/fault removal,
- `morph.cpp` now gates `setFramePointerRequiredEH` on `compHndBBtabCount` (live count) instead of `info.compXcptnsCount` (IL count),
- the `EHSlots` / `ehNeedsShadowSPslots` machinery is gone entirely after #115957 (new x86 EH).
